arduino/examples/ESP32CanLoadTest/ESP32CanLoadTest.ino

```diff
@@ -0,0 +1,401 @@
+
+
+// Create an initializer that can initialize all the GPIO pins in one shot
+typedef GpioInitializer<
+    IO0_Pin,  IO1_Pin,  IO2_Pin,  IO3_Pin,  // outputs 0-3
+    IO4_Pin,  IO5_Pin,  IO6_Pin,  IO7_Pin,  // outputs 4-7
+    IO8_Pin,  IO9_Pin,  IO10_Pin, IO11_Pin, // inputs 0-3
+    IO12_Pin, IO13_Pin, IO14_Pin, IO15_Pin // inputs 4-7
+    > GpioInit;
+
+// The producers need to be polled repeatedly for changes and to execute the
+// debouncing algorithm. This class instantiates a refreshloop and adds the
+// producers to it.
+openlcb::RefreshLoop producer_refresh_loop(openmrn.stack()->node(),
+    {
+        IO8_producer.polling(),
+        IO9_producer.polling(),
+        IO10_producer.polling(),
+        IO11_producer.polling(),
+        IO12_producer.polling(),
+        IO13_producer.polling(),
+        IO14_producer.polling(),
+        IO15_producer.polling()
+    }
+);
+
+class FactoryResetHelper : public DefaultConfigUpdateListener {
+public:
+    UpdateAction apply_configuration(int fd, bool initial_load,
+                                     BarrierNotifiable *done) OVERRIDE {
+        AutoNotify n(done);
+        return UPDATED;
+    }
+
+    void factory_reset(int fd) override
+    {
+        cfg.userinfo().name().write(fd, openlcb::SNIP_STATIC_DATA.model_name);
+        cfg.userinfo().description().write(
+            fd, "OpenLCB + Arduino-ESP32 on an ESP32.");
+        for(int i = 0; i < openlcb::NUM_OUTPUTS; i++)
+        {
+            cfg.seg().consumers().entry(i).description().write(fd, "");
+        }
+        for(int i = 0; i < openlcb::NUM_INPUTS; i++)
+        {
+            cfg.seg().producers().entry(i).description().write(fd, "");
+            CDI_FACTORY_RESET(cfg.seg().producers().entry(i).debounce);
+        }
+    }
+} factory_reset_helper;
+
+namespace openlcb
+{
+    // Name of CDI.xml to generate dynamically.
+    const char CDI_FILENAME[] = "/spiffs/cdi.xml";
+
+    // This will stop openlcb from exporting the CDI memory space upon start.
+    extern const char CDI_DATA[] = "";
+
+    // Path to where OpenMRN should persist general configuration data.
+    extern const char *const CONFIG_FILENAME = "/spiffs/openlcb_config";
+
+    // The size of the memory space to export over the above device.
+    extern const size_t CONFIG_FILE_SIZE = cfg.seg().size() + cfg.seg().offset();
+
+    // Default to store the dynamic SNIP data is stored in the same persistant
+    // data file as general configuration data.
+    extern const char *const SNIP_DYNAMIC_FILENAME = CONFIG_FILENAME;
+}
+
+CpuLoad cpu_load;
+hw_timer_t * timer = nullptr;
+CpuLoadLog* cpu_log = nullptr;
+
+void IRAM_ATTR onTimer()
+{
+  if (spi_flash_cache_enabled()) {
+    cpuload_tick(0);
+  }
+}
+
+void setup()
+{
+    Serial.begin(115200L);
+
+    timer = timerBegin(3, 80, true); // timer_id = 3; divider=80; countUp = true;
+    timerAttachInterrupt(timer, &onTimer, true); // edge = true
+    timerAlarmWrite(timer, 1000000/163, true);  //1000 ms
```

Comment:
comment does not reflect intention here. We can increase the divider from 80 to 80000 and replace this math with 1000/163 (or ~6).

Reply:
fixed.
1000/6 is not good because that will run the timer in sync with the kernel ticks. We need the two timers to be running at a relative prime interval to avoid systematic error in the sampling.

Reply:
I wasn't meaning 1000/6 but instead 6uS by itself which is close to the 1M/163 that is used now. The updated commit though clarifies the intent enough though.